src/notification/NotificationController.ts

```diff
@@ -0,0 +1,215 @@
+
+export type NotificationControllerMessenger = RestrictedControllerMessenger<
+  typeof name,
+  ControllerActions,
+  NotificationControllerStateChange,
+  never,
+  never
+>;
+
+const metadata = {
+  notifications: { persist: true, anonymous: false },
+};
+
+const defaultState = {
+  notifications: {},
+};
+
+/**
+ * Controller that handles storing notifications and showing them to the user
+ */
+export class NotificationController extends BaseController<
+  typeof name,
+  NotificationControllerState,
+  NotificationControllerMessenger
+> {
+  private showNativeNotification;
+
+  /**
+   * Creates a NotificationController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.showNativeNotification - Function that shows a native notification in the consumer
+   */
+  constructor({
+    messenger,
+    state,
+    showNativeNotification,
+  }: {
+    messenger: NotificationControllerMessenger;
+    state?: Partial<NotificationControllerState>;
+    showNativeNotification: (origin: string, message: string) => void;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.showNativeNotification = showNativeNotification;
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:show` as const,
+      (origin: string, args: NotificationArgs) => this.show(origin, args),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:dismiss` as const,
+      (ids: string[]) => this.dismiss(ids),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:markRead` as const,
+      (ids: string[]) => this.markRead(ids),
+    );
+  }
+
+  /**
+   * Shows a notification.
+   *
+   * @param origin - The origin trying to send a notification
+   * @param args - Notification arguments, containing the notification message etc.
+   */
+  show(origin: string, args: NotificationArgs) {
+    switch (args.type) {
+      case NotificationType.Native:
+        this.showNativeNotification(origin, args.message);
+        break;
+      case NotificationType.InApp:
+        this.add(origin, args.message);
+        break;
+      default:
+        throw new Error('Invalid notification type');
+    }
+  }
+
+  private add(origin: string, message: string) {
+    const id = nanoid();
+    const notification = {
+      id,
+      type: NotificationType.InApp,
+      origin,
+      createdDate: Date.now(),
+      readDate: null,
+      message,
+    };
+    this.update((state) => {
+      state.notifications[id] = notification;
+    });
+  }
+
+  /**
+   * Dimisses a list of notifications.
+   *
+   * @param ids - A list of notification IDs
+   */
+  dismiss(ids: string[]) {
+    this.update((state) => {
+      for (const id of ids) {
+        if (hasProperty(state.notifications, id)) {
```

Comment:
Would this suffice?

``` suggestion
        if (id in state.notifications) {
```

Reply:
The `in` keyword looks up the entire prototype chain, which isn't typically what people expect it to do. Erik has proposed that we disallow it entirely for this reason (see https://github.com/MetaMask/eslint-config/pull/207)